core/src/main/java/org/infinispan/util/concurrent/locks/impl/InfinispanLock.java

```diff
@@ -0,0 +1,537 @@
+      Objects.requireNonNull(timeUnit, "Time Unit should be non-null");
+
+      if (trace) {
+         log.tracef("Acquire lock for %s. Timeout=%s (%s)", lockOwner, time, timeUnit);
+      }
+
+      LockPlaceHolder lockPlaceHolder = lockOwners.get(lockOwner);
+      if (lockPlaceHolder != null) {
+         if (trace) {
+            log.tracef("Lock owner already exists: %s", lockPlaceHolder);
+         }
+         return lockPlaceHolder;
+      }
+
+      lockPlaceHolder = createLockInfo(lockOwner, time, timeUnit);
+      LockPlaceHolder other = lockOwners.putIfAbsent(lockOwner, lockPlaceHolder);
+
+      if (other != null) {
+         if (trace) {
+            log.tracef("Lock owner already exists: %s", other);
+         }
+         return other;
+      }
+
+      if (trace) {
+         log.tracef("Created a new one: %s", lockPlaceHolder);
+      }
+
+      pendingRequest.add(lockPlaceHolder);
+      tryAcquire();
+      return lockPlaceHolder;
+   }
+
+   /**
+    * It tries to release the lock held by {@code lockOwner}.
+    * <p/>
+    * If the lock is not acquired (is waiting or timed out/deadlocked) by {@code lockOwner}, its {@link
+    * ExtendedLockPromise} is canceled. If {@code lockOwner} is the current lock owner, the lock is released and the
+    * next lock owner available will acquire the lock. If the {@code lockOwner} never tried to acquire the lock, this
+    * method does nothing.
+    *
+    * @param lockOwner the lock owner who wants to release the lock.
+    * @throws NullPointerException if {@code lockOwner} is {@code null}.
+    */
+   public void release(Object lockOwner) {
+      Objects.requireNonNull(lockOwner, "Lock Owner should be non-null");
+
+      if (trace) {
+         log.tracef("Release lock for %s.", lockOwner);
+      }
+
+      LockPlaceHolder wantToRelease = lockOwners.get(lockOwner);
+      if (wantToRelease == null) {
+         if (trace) {
+            log.tracef("%s not found!", lockOwner);
+         }
+         //nothing to release
+         return;
+      }
+
+      final boolean released = wantToRelease.setReleased();
+      if (trace) {
+         log.tracef("Release lock for %s? %s", wantToRelease, released);
+      }
+
+      LockPlaceHolder currentLocked = current;
+      if (currentLocked == wantToRelease) {
+         if (!casRelease(currentLocked)) {
+            if (trace) {
+               log.tracef("Releasing lock for %s. It is the current lock owner but another thread changed it.", lockOwner);
+            }
+            //another thread already released!
+         } else {
+            tryAcquire();
+         }
+      }
+   }
+
+   /**
+    * @return {@code true} if this lock has no lock owner pending.
+    */
+   public boolean isEmpty() {
+      return isFree() && lockOwners.isEmpty();
+   }
+
+   /**
+    * @return {@code true} if the lock is not acquired and no lock owners are queued.
+    */
+   public boolean isFree() {
```

Comment:
I think `isFree()` and `isEmpty()` could both be replaced with `!isLocked()`, because it shouldn't be possible to have queued requests without a current owner.